walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,123 @@
+  * a meeting (with duration and mandatory/optional attendees), return the times when the meeting 
+  * could happen that day.
+**/
 public final class FindMeetingQuery {
+    private static final TimeRange START_OF_DAY_TIMERANGE =
+        TimeRange.fromStartDuration(TimeRange.START_OF_DAY, 0);
+    private static final TimeRange END_OF_DAY_TIMERANGE =
+        TimeRange.fromStartDuration(TimeRange.END_OF_DAY + 1, 0);
+
+  /** returns time ranges that will accomodate meeting request
+   *  if no time ranges exist all (mandatory + optional) attendees can attend, time ranges that
+   *  accomodate only mandatory attendees is returned instead
+  **/
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> allAttendeeTimeRanges = extractAttendingTimeRangesFromEvents(
+        events, request, /* include optional attendees= */ true);
+    Collection<TimeRange> acceptableOptionalTimeRanges = getConflictFreeTimeRanges(
+        allAttendeeTimeRanges, request);
+
+    if (request.getAttendees().isEmpty() || !acceptableOptionalTimeRanges.isEmpty()) {
+      return acceptableOptionalTimeRanges;
+    } else {
+      Collection<TimeRange> mandatoryAttendeeTimeRanges = extractAttendingTimeRangesFromEvents(
+          events, request, /* includes optional attendees= */ false);
+      Collection<TimeRange> acceptableMandatoryTimeRanges = getConflictFreeTimeRanges(
+          mandatoryAttendeeTimeRanges, request);
+      return acceptableMandatoryTimeRanges;
+    }
+  }
+  
+  /** given a collection of TimeRanges and a meeting request, returns 
+    * a collection of timeranges that the meeting can be scheduled in 
+    * without conflicts
+  **/
+  private Collection<TimeRange> getConflictFreeTimeRanges(
+      Collection<TimeRange> timeRanges, MeetingRequest request) {
+        
+    // edge case: if meeting request duration is too long, return empty list
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // edge case: if there are no exisiting meetings, return entire day
+    if (timeRanges.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    List<TimeRange> timeRangesByStart = prepareAndSortTimeRanges(timeRanges);
+    
+    List<TimeRange> acceptableTimeRanges = new ArrayList<TimeRange>();
+    TimeRange currentMeetingTimeRange;
+    TimeRange nextMeetingTimeRange;
+
+    for (int i = 0; i < timeRangesByStart.size() - 1; i++) {
+      currentMeetingTimeRange = timeRangesByStart.get(i);  
+      nextMeetingTimeRange = timeRangesByStart.get(i + 1);
+      // if overlap, merge with next meeting
+      if (currentMeetingTimeRange.overlaps(nextMeetingTimeRange)) {
+        timeRangesByStart.set(
+            i + 1, getMergedMeetingTimeRange(currentMeetingTimeRange, nextMeetingTimeRange));
```

Comment:
Why are we replacing `i + 1`?

Reply:
I was trying to only do a single pass over the time-ranges. The merge and replace of `i+1` happens so that if the current time-range overlaps with the next time-range, the next time-range "absorbs" the current one and the current one is skipped.